src/rebase.rs

```diff
@@ -0,0 +1,487 @@
+    /// the index and working directory will be updated with the changes. If
+    /// there are conflicts, you will need to address those before committing
+    /// the changes.
+    ///
+    /// Iterating over the result of `operations()` correctly handles begin and
+    /// end conditions for you.
+    pub fn next<'rebase>(&'rebase self) -> Result<RebaseOperation<'rebase, 'repo>, Error> {
+        let mut ret = 0 as *mut raw::git_rebase_operation;
+        unsafe {
+            try_call!(raw::git_rebase_next(&mut ret, self.raw));
+            Ok(Binding::from_raw(ret))
+        }
+    }
+
+    /// Iterator of rebase operations
+    pub fn operations<'rebase>(&'rebase self) -> RebaseOperations<'rebase, 'repo> {
+        RebaseOperations {
+            rebase: self,
+            operation_count: self.operation_count(),
+        }
+    }
+
+    /// Gets the rebase operation specified by the given index.
+    pub fn operation_at_index<'rebase>(
+        &'rebase self,
+        index: usize,
+    ) -> Option<RebaseOperation<'rebase, 'repo>> {
+        unsafe {
+            let ptr = raw::git_rebase_operation_byindex(self.raw, index as size_t);
+            if ptr.is_null() {
+                None
+            } else {
+                Some(Binding::from_raw(ptr))
+            }
+        }
+    }
+
+    /// Gets the index of the rebase operation that is currently being applied.
+    ///
+    /// If the first operation has not yet been applied it returns `None`.
+    pub fn current_operation_index(&self) -> Option<usize> {
+        unsafe {
+            let index = raw::git_rebase_operation_current(self.raw);
+            if index == raw::GIT_REBASE_NO_OPERATION {
+                None
+            } else {
+                Some(index as usize)
+            }
+        }
+    }
+
+    /// Convenience function to get the operation at the current index.
+    ///
+    /// Returns `None` if the first operation has not yet been applied.
+    pub fn current_operation<'rebase>(&'rebase self) -> Option<RebaseOperation<'rebase, 'repo>> {
+        self.current_operation_index()
+            .and_then(|index| self.operation_at_index(index))
+    }
+}
+
+impl<'rebase, 'repo: 'rebase> Iterator for RebaseOperations<'rebase, 'repo> {
+    type Item = Result<RebaseOperation<'rebase, 'repo>, Error>;
+    /// Return the next rebase operation that needs to be performed
+    ///
+    /// Verifies that there _is_ a next operation and then calls
+    /// `Rebase::next`, propagating any errors from there.
+    fn next(&mut self) -> Option<Result<RebaseOperation<'rebase, 'repo>, Error>> {
+        // The current_operation_index() returns the value for the _previous_
+        // invocation of rebase.next(), which means it is None on the first
+        // invocation and we need to offset idx + 1 for the general "are we
+        // done with the loop" query.
+        match self.rebase.current_operation_index() {
+            None => Some(self.rebase.next()),
```

Comment:
I think the usage of `next()` here will cause the iterator to not work when run twice, so perhaps this could use `operation_at_index`?

Reply:
I think that each `RebaseOprations` needs to be considered a 1-shot iterator, because of the state that git keeps internally.